packages/server/src/lockers/GCSLocker.ts

```diff
@@ -0,0 +1,155 @@
+   */
+  lockTTL?: number
+  /**
+   * The amount of time (in milliseconds) to wait between lock file health checks. Must be set according to `lockTTL`, and must be less than `acquireLockTimeout`. Larger value results less queries to GCS.
+   */
+  watchInterval?: number
+}
+
+export class GCSLocker implements Locker {
+  events: EventEmitter
+  bucket: Bucket
+  lockTimeout: number
+  lockTTL: number
+  watchInterval: number
+
+  constructor(options: GCSLockerOptions) {
+    this.events = new EventEmitter()
+    this.bucket = options.bucket
+    this.lockTimeout = options.acquireLockTimeout ?? 1000 * 30
+    this.lockTTL = options.lockTTL ?? 1000 * 12
+    this.watchInterval = options.watchInterval ?? 1000 * 10
+
+    if (this.watchInterval > this.lockTimeout) {
+      throw new Error('watchInterval must be less than acquireLockTimeout')
+    }
+  }
+
+  newLock(id: string) {
+    return new GCSLockHandler(id, this)
+  }
+}
+
+class GCSLockHandler implements Lock {
+  private gcsLock: GCSLock
+
+  constructor(private id: string, private locker: GCSLocker) {
+    this.gcsLock = new GCSLock(
+      this.id,
+      this.locker.bucket,
+      this.locker.lockTTL,
+      this.locker.watchInterval
+    )
+  }
+
+  async lock(requestRelease: RequestRelease): Promise<void> {
+    const abortController = new AbortController()
+
+    const lock = await Promise.race([
+      this.waitForLockTimeoutOrAbort(abortController.signal),
+      this.acquireLock(requestRelease, abortController.signal),
+    ])
+
+    abortController.abort()
+
+    if (!lock) {
+      throw ERRORS.ERR_LOCK_TIMEOUT
+    }
+  }
+
+  async unlock(): Promise<void> {
+    await this.gcsLock.release()
+  }
+
+  protected async acquireLock(
+    cancelHandler: RequestRelease,
+    signal: AbortSignal,
+    attempt = 0
+  ): Promise<boolean> {
+    if (signal.aborted) {
+      return false
+    }
+
+    const acquired = await this.gcsLock.take(cancelHandler)
+
+    if (!acquired) {
+      //Try to acquire the lock again
+      return await new Promise((resolve, reject) => {
+        //On the first attempt, retry after current I/O operations are done, else use an exponential backoff
+        const waitFn = (then: () => void) =>
+          attempt > 0
+            ? setTimeout(then, (attempt * this.locker.watchInterval) / 3)
+            : setImmediate(then)
+
+        waitFn(() => {
+          this.acquireLock(cancelHandler, signal, attempt + 1)
+            .then(resolve)
+            .catch(reject)
+        })
+      })
+    }
+
+    return true
+  }
+
+  protected waitForLockTimeoutOrAbort(signal: AbortSignal) {
+    return new Promise<boolean>((resolve) => {
```

Comment:
Node.js has a built-in async version of setTimeout: https://nodejs.org/api/timers.html#timerspromisessettimeoutdelay-value-options